File: android/app/src/main/kotlin/run/nuri/getagrip/ble/BroadcastGaugeClient.kt
// SPDX-License-Identifier: MPL-2.0 AND BSD-2-Clause
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ble

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import run.nuri.getagrip.engine.ForceSample
import run.nuri.getagrip.engine.GaugeKind
import run.nuri.getagrip.engine.L10n
import run.nuri.getagrip.engine.ProgressorCommand
import run.nuri.getagrip.engine.ProgressorEvent
import run.nuri.getagrip.engine.SyntheticSampleClock
import run.nuri.getagrip.engine.WHC06Codec
import kotlin.math.ceil
import kotlin.math.max

/// WH-C06 sends weight in advertisements: no GATT link and nothing to write. Ordinary
/// gaps, tare and stream starts leave its healthy scan alone. Ten seconds of real silence
/// retires the link and replaces its scan, with bounded retries rather than waiting forever
/// on a wedged scanner. Each registration has a fresh listener; a link owns its address
/// lock, tare and watchdog. The scan quota survives disconnect/reconnect and radio-off.
///
/// Protocol knowledge ported from hangtime-grip-connect (BSD-2-Clause, © 2024
/// Stevie-Ray Hartog, https://github.com/Stevie-Ray/hangtime-grip-connect).
class BroadcastGaugeClient internal constructor(
    private val transport: BroadcastScanTransport,
    private val scope: CoroutineScope,
    private val clock: HostClock = SystemHostClock,
) : ProgressorClient {
    constructor(context: Context, scope: CoroutineScope, clock: HostClock = SystemHostClock) :
        this(AndroidBroadcastScanTransport(context), scope, clock)

    override var onEvent: ((ProgressorEvent) -> Unit)? = null
    override var onPacketBoundary: ((PacketBoundary) -> Unit)? = null
    override var onStateChange: ((ProgressorConnectionState) -> Unit)? = null
    override var onDiagnostic: ((ProgressorClientDiagnostic) -> Unit)? = null
    override var state: ProgressorConnectionState = ProgressorConnectionState.Idle
        private set(value) {
            val changed = field != value
            field = value
            if (changed) onStateChange?.invoke(value)
        }
    override var deviceName: String? = null
        private set
    override val kind: GaugeKind = GaugeKind.whc06

    internal companion object {
        const val firstFrameDeadlineMillis = 15_000L
        const val silencePollMillis = 1_000L
        const val maximumAcquisitionAttempts = 3
        // Below Android's five-start limit. An extra second avoids boundary rounding races.
        const val scanRateWindowSeconds = 31.0
        const val maximumScanStartsPerWindow = 4
        const val scanTooFrequentError = 6
    }

    private var wantsConnection = false
    private var generation = 0uL
    private var activeScan: BroadcastScanTransport.Listener? = null
    private var lockedAddress: String? = null
    private var lastFrameUptime: Double? = null
    private var lastAdvertisementUptime: Double? = null
    private val softwareTare = SoftwareTare()
    private var silenceJob: Job? = null
    private var firstFrameJob: Job? = null
    private var retryJob: Job? = null
    private var acquisitionAttempts = 0
    private var acquisitionReason = "initial"
    private val recentScanStarts = ArrayDeque<Double>()
    private var scanCooldownUntil = 0.0

    override fun connect() {
        if (wantsConnection || state.isBusy || state.isConnected) return
        wantsConnection = true
        transport.observeRadio { onMain { radioChanged() } }
        if (!radioAvailable()) return
        beginAcquisition("initial")
    }

    override fun disconnect() {
        wantsConnection = false
        generation += 1uL
        cancelAllJobs()
        retireScan("disconnect")
        resetLink()
        transport.stopObservingRadio()
        state = ProgressorConnectionState.Disconnected()
    }

    override fun sleepDevice() = disconnect()

    override fun send(command: ProgressorCommand) {
        when (command) {
            ProgressorCommand.tare -> softwareTare.capture()
            ProgressorCommand.enterSleep -> disconnect()
            // Stopping measurement does not stop its scan/link. The store decides which
            // readings to record. WH-C06 has no hardware commands of any kind.
            else -> Unit
        }
    }

    override fun startStreaming(cause: StreamStartCause) {
        if (!wantsConnection) {
            diagnostic("scan not requested (${cause.label}, disconnected)")
            return
        }
        if (!radioAvailable()) {
            diagnostic("scan unavailable (${cause.label}, radio unavailable)")
            return
        }
        when {
            activeScan != null -> diagnostic("scan already active (${cause.label})")
            retryJob != null -> diagnostic("scan retry pending (${cause.label})")
            else -> requestScan()
        }
        // No tare-deferred/write diagnostics: this gauge has no command surface. Even startScan's
        // return does not confirm the platform's asynchronous registration succeeded.
    }

    private fun beginAcquisition(reason: String) {
        acquisitionAttempts = 0
        acquisitionReason = reason
        state = ProgressorConnectionState.Scanning
        requestScan()
    }

    private fun requestScan(minimumDelaySeconds: Double = 0.0) {
        if (!wantsConnection || activeScan != null || !radioAvailable()) return
        retryJob?.cancel()
        retryJob = null
        val now = clock.uptimeSeconds()
        while (recentScanStarts.isNotEmpty() && now - recentScanStarts.first() >= scanRateWindowSeconds) {
            recentScanStarts.removeFirst()
        }
        val quotaDelay = if (recentScanStarts.size >= maximumScanStartsPerWindow) {
            recentScanStarts.first() + scanRateWindowSeconds - now
        } else 0.0
        val waitSeconds = max(minimumDelaySeconds, max(quotaDelay, scanCooldownUntil - now))
        if (waitSeconds > 0) {
            val watchGeneration = generation
            diagnostic("scan retry in ${ceil(waitSeconds).toInt()} s")
            retryJob = scope.launch(Dispatchers.Main.immediate) {
                delay(ceil(waitSeconds * 1_000).toLong())
                if (generation != watchGeneration || !wantsConnection) return@launch
                retryJob = null
                requestScan()
            }
            return
        }
        startScanAttempt(now)
    }

    private fun startScanAttempt(startedAt: Double) {
        if (acquisitionAttempts >= maximumAcquisitionAttempts) {
            finishFailure("scan recovery exhausted")
            return
        }
        acquisitionAttempts += 1
        recentScanStarts.addLast(startedAt)
        val listener = object : BroadcastScanTransport.Listener {
            override fun onAdvertisement(advertisement: BroadcastAdvertisement) {
                onMain { if (activeScan === this) didAdvertise(advertisement, startedAt) }
            }
            override fun onFailure(errorCode: Int) {
                onMain { if (activeScan === this) scanFailed(errorCode) }
            }
        }
        activeScan = listener
        diagnostic("scan requested ($acquisitionReason, attempt $acquisitionAttempts)")
        try {
            val started = transport.start(listener)
            // A synchronous failure/result is permitted; never arm a retired attempt.
            if (activeScan !== listener) return
            if (!started) {
                attemptFailed("scanner unavailable")
                return
            }
        } catch (_: SecurityException) {
            finishFailure("scan permission denied", ProgressorConnectionState.Unauthorized)
            return
        } catch (_: IllegalStateException) {
            attemptFailed("scanner unavailable")
            return
        }
        if (!state.isConnected) startFirstFrameDeadline(listener)
    }

    private fun scanFailed(errorCode: Int) {
        if (errorCode == scanTooFrequentError) {
            scanCooldownUntil = max(scanCooldownUntil, clock.uptimeSeconds() + scanRateWindowSeconds)
        }
        attemptFailed("scan failed (code $errorCode)")
    }

    private fun attemptFailed(reason: String) {
        diagnostic(reason)
        firstFrameJob?.cancel()
        firstFrameJob = null
        retireScan("failed attempt")
        if (state.isConnected) {
            // onScanFailed normally precedes results. If sent later, publish a real link
            // break before retrying so the runner cannot accrue across the missing link.
            generation += 1uL
            cancelAllJobs()
            resetLink()
            state = ProgressorConnectionState.Disconnected(L10n.tr("Scale stopped broadcasting"))
            if (!wantsConnection) return
            acquisitionAttempts = 0
            acquisitionReason = "recovery"
        }
        if (acquisitionAttempts >= maximumAcquisitionAttempts) {
            finishFailure("scan recovery exhausted")
            return
        }
        state = ProgressorConnectionState.Scanning
        requestScan(if (acquisitionAttempts <= 1) 2.0 else 5.0)
    }

    private fun startFirstFrameDeadline(listener: BroadcastScanTransport.Listener) {
        firstFrameJob?.cancel()
        firstFrameJob = scope.launch(Dispatchers.Main.immediate) {
            delay(firstFrameDeadlineMillis)
            if (activeScan !== listener || state.isConnected || !wantsConnection) return@launch
            firstFrameJob = null
            attemptFailed("scan timed out")
        }
    }

    private fun didAdvertise(advertisement: BroadcastAdvertisement, scanStartedAt: Double) {
        if (!wantsConnection || !radioAvailable()) return
        val receivingScan = activeScan
        // Listener identity rejects retired callbacks; native receipt time additionally
        // rejects buffered pre-scan results delivered through the NEW callback.
        val observedAt = advertisement.observedUptimeSeconds
        if (!observedAt.isFinite() || observedAt < scanStartedAt) return
        if (lastAdvertisementUptime?.let { observedAt <= it } == true) return
        val payload = advertisement.manufacturerPayload
        val framed = ByteArray(payload.size + 2)
        framed[0] = (WHC06Codec.companyID and 0xFF).toByte()
        framed[1] = ((WHC06Codec.companyID shr 8) and 0xFF).toByte()
        payload.copyInto(framed, 2)
        val rawKg = WHC06Codec.kilogramsFromManufacturerData(framed) ?: return
        if (lockedAddress?.let { it != advertisement.address } == true) return
        lockedAddress = advertisement.address
        lastAdvertisementUptime = observedAt
        if (advertisement.name != null) deviceName = advertisement.name
        val uptime = clock.uptimeSeconds()
        lastFrameUptime = uptime
        if (!state.isConnected) {
            firstFrameJob?.cancel()
            firstFrameJob = null
            acquisitionAttempts = 0
            diagnostic("scale advertisements received")
            state = ProgressorConnectionState.Connected
        }
        // State observers may immediately disconnect (for example on backgrounding).
        // Do not publish the just-arrived sample after that decision ended its link.
        if (!wantsConnection || activeScan !== receivingScan) return
        startSilenceWatchdog()
        val kg = softwareTare.value(rawKg)
        withPacket(uptime) {
            onEvent?.invoke(ProgressorEvent.Sample(
                ForceSample(kg, SyntheticSampleClock.micros(uptime), isBatchStart = true),
            ))
        }
    }

    private fun startSilenceWatchdog() {
        if (silenceJob != null) return
        val watchGeneration = generation
        silenceJob = scope.launch(Dispatchers.Main.immediate) {
            while (isActive) {
                delay(silencePollMillis)
                if (generation != watchGeneration || !state.isConnected) return@launch
                val last = lastFrameUptime ?: continue
                if (clock.uptimeSeconds() - last < WHC06Codec.advertisementSilenceSeconds) continue
                silenceJob = null
                handleSilence()
                return@launch
            }
        }
    }

    private fun handleSilence() {
        generation += 1uL
        cancelAllJobs()
        retireScan("advertisement silence")
        resetLink()
        state = ProgressorConnectionState.Disconnected(L10n.tr("Scale stopped broadcasting"))
        if (wantsConnection) beginAcquisition("recovery")
    }

    private fun retireScan(reason: String) {
        val previous = activeScan ?: return
        activeScan = null
        runCatching { transport.stop(previous) }
        diagnostic("scan stopped ($reason)")
    }

    private fun resetLink() {
        lockedAddress = null
        lastFrameUptime = null
        lastAdvertisementUptime = null
        deviceName = null
        softwareTare.reset()
    }

    private fun cancelAllJobs() {
        silenceJob?.cancel()
        silenceJob = null
        firstFrameJob?.cancel()
        firstFrameJob = null
        retryJob?.cancel()
        retryJob = null
    }

    private fun finishFailure(
        reason: String,
        terminalState: ProgressorConnectionState = ProgressorConnectionState.Disconnected(L10n.tr("No scale found")),
    ) {
        diagnostic(reason)
        wantsConnection = false
        generation += 1uL
        cancelAllJobs()
        retireScan("giving up")
        resetLink()
        transport.stopObservingRadio()
        state = terminalState
    }

    private fun radioAvailable(): Boolean {
        val radio = transport.radioState
        if (radio == ProgressorConnectionState.Idle) return true
        if (radio == ProgressorConnectionState.Unauthorized || radio == ProgressorConnectionState.Unsupported) {
            finishFailure("radio unavailable", radio)
        } else {
            generation += 1uL
            cancelAllJobs()
            retireScan("radio unavailable")
            resetLink()
            state = radio
        }
        return false
    }

    private fun radioChanged() {
        if (!wantsConnection || !radioAvailable()) return
        if (activeScan == null && retryJob == null) beginAcquisition("radio restored")
    }

    private fun diagnostic(event: String) {
        onDiagnostic?.invoke(ProgressorClientDiagnostic.BroadcastScan(event))
    }

    private fun onMain(block: () -> Unit) {
        scope.launch(Dispatchers.Main.immediate) { block() }
    }
}

File: Sources/BLE/BroadcastGaugeClient.swift
// SPDX-License-Identifier: MPL-2.0 AND BSD-2-Clause
// Original contributions Copyright 2026 Nuri Bruner.

import Foundation

/// WH-C06 measurements are advertisements, not a connected GATT stream. A healthy
/// scan remains running; only a lost signal starts bounded acquisition attempts.
/// Protocol knowledge ported from hangtime-grip-connect (BSD-2-Clause, © 2024
/// Stevie-Ray Hartog, https://github.com/Stevie-Ray/hangtime-grip-connect).
@MainActor
final class BroadcastGaugeClient: ProgressorClient {
    var onEvent: ((ProgressorEvent) -> Void)?
    var onPacketBoundary: ((PacketBoundary) -> Void)?
    var onStateChange: ((ProgressorConnectionState) -> Void)?
    var onDiagnostic: ((ProgressorClientDiagnostic) -> Void)?

    private(set) var state: ProgressorConnectionState = .idle {
        didSet { if state != oldValue { onStateChange?(state) } }
    }
    private(set) var deviceName: String?
    let kind: GaugeKind = .whc06

    private enum Acquisition: String {
        case initial
        case recovery
        var maximumAttempts: Int { self == .initial ? 1 : 3 }
    }

    private static let firstFrameDeadlineSeconds: TimeInterval = 15
    private static let retryDelays: [TimeInterval] = [2, 5]

    private let transport: any BroadcastScanTransport
    private let scheduler: any BroadcastScanScheduler
    private var wantsConnection = false
    private var isScanning = false
    private var scanGeneration: UInt64 = 0
    private var acquisition = Acquisition.initial
    private var attempt = 0

    /// First valid advertiser owns this logical link. Another scale must never
    /// interleave its weights or inherit a software zero captured against this one.
    private var lockedPeripheral: UUID?
    private var softwareTare = SoftwareTare()
    private var lastFrameUptime: TimeInterval?
    private var firstFrameTimer: (any BroadcastScanCancellation)?
    private var silenceTimer: (any BroadcastScanCancellation)?
    private var retryTimer: (any BroadcastScanCancellation)?

    convenience init() {
        self.init(transport: CoreBluetoothBroadcastScanTransport(),
                  scheduler: MonotonicBroadcastScanScheduler())
    }

    /// The real lifecycle is testable without creating CoreBluetooth (which does
    /// not exist in the Simulator) or waiting through radio timeouts in tests.
    init(transport: any BroadcastScanTransport, scheduler: any BroadcastScanScheduler) {
        self.transport = transport
        self.scheduler = scheduler
        transport.onRadioStateChange = { [weak self] state in
            self?.radioStateChanged(state)
        }
    }

    // MARK: - ProgressorClient

    func connect() {
        guard !wantsConnection, !state.isBusy, !state.isConnected else { return }
        wantsConnection = true
        acquisition = .initial
        attempt = 0
        // Constructing the central raises the Bluetooth permission prompt. Keep
        // that construction behind an explicit Connect, including with this seam.
        transport.activate()
        if transport.radioState == .poweredOn {
            beginScan()
        } else {
            state = Self.connectionState(for: transport.radioState)
        }
    }

    func disconnect() {
        wantsConnection = false
        cancelRetry()
        retireScan(reason: "disconnect")
        clearLink()
        state = .disconnected(reason: nil)
    }

    func sleepDevice() { disconnect() }

    func send(_ command: ProgressorCommand) {
        switch command {
        case .tare:
            softwareTare.capture()
        case .enterSleep:
            disconnect()
        default:
            // There is no command characteristic. Stop only changes recording
            // in DeviceStore; the scale keeps broadcasting and the scan is its link.
            break
        }
    }

    func startStreaming(cause: StreamStartCause) {
        guard wantsConnection, transport.radioState == .poweredOn else {
            diagnostic("scan unavailable (\(cause.label))")
            return
        }
        if isScanning {
            // Ordinary advertisement gaps and tare must not bounce a healthy scan.
            diagnostic("scan already active (\(cause.label))")
        } else if retryTimer != nil {
            diagnostic("scan recovery pending (\(cause.label))")
        } else {
            beginScan()
        }
    }

    // MARK: - Acquisition

    private func beginScan() {
        guard wantsConnection, transport.radioState == .poweredOn,
              !isScanning, retryTimer == nil else { return }
        attempt += 1
        scanGeneration &+= 1
        let generation = scanGeneration
        isScanning = true
        state = .scanning
        // A state observer may cancel synchronously; never resurrect that intent.
        guard ownsScan(generation) else { return }
        diagnostic("scan requested (\(acquisition.rawValue), attempt \(attempt))")
        firstFrameTimer = scheduler.schedule(after: Self.firstFrameDeadlineSeconds) { [weak self] in
            guard let self, self.ownsScan(generation), self.state == .scanning else { return }
            self.diagnostic("scan timed out")
            self.acquisitionFailed()
        }
        let started = transport.startScan { [weak self] advertisement in
            self?.received(advertisement, generation: generation)
        }
        if !started, ownsScan(generation) {
            diagnostic("scan could not start")
            acquisitionFailed()
        }
    }

    private func acquisitionFailed() {
        retireScan(reason: "failed attempt")
        guard wantsConnection, transport.radioState == .poweredOn else { return }
        guard attempt < acquisition.maximumAttempts else {
            wantsConnection = false
            clearLink()
            if acquisition == .recovery { diagnostic("scan recovery exhausted") }
            state = .disconnected(reason: String(localized: "No scale found"))
            return
        }

        let delay = Self.retryDelays[attempt - 1]
        let generation = scanGeneration
        diagnostic("scan retry in \(Int(delay)) s")
        retryTimer = scheduler.schedule(after: delay) { [weak self] in
            guard let self, self.wantsConnection, self.scanGeneration == generation,
                  self.transport.radioState == .poweredOn else { return }
            self.retryTimer = nil
            self.beginScan()
        }
    }

    private func received(_ advertisement: BroadcastAdvertisement, generation: UInt64) {
        guard ownsScan(generation),
              let rawKg = WHC06Codec.kilograms(fromManufacturerData: advertisement.manufacturerData),
              lockedPeripheral == nil || lockedPeripheral == advertisement.peripheralID else { return }

        lockedPeripheral = advertisement.peripheralID
        if let name = advertisement.name { deviceName = name }
        let uptime = scheduler.now
        lastFrameUptime = uptime
        if !state.isConnected {
            firstFrameTimer?.cancel()
            firstFrameTimer = nil
            diagnostic("scale advertisements received")
            state = .connected
        }
        guard ownsScan(generation) else { return }
        armSilenceTimer(generation: generation)

        // iOS supplies no public native observation timestamp. This is the
        // monotonic CALLBACK time, matching the synthetic-clock contract; it does
        // not claim that a radio-delayed packet was measured at that instant.
        let sample = ForceSample(kg: softwareTare.value(for: rawKg),
                                 deviceMicros: SyntheticSampleClock.micros(uptime: uptime),
                                 isBatchStart: true)
        onPacketBoundary?(.began(receivedAt: uptime))
        defer { onPacketBoundary?(.ended) }
        onEvent?(.sample(sample))
    }

    /// One timer follows the last reading; accepting a packet never allocates or
    /// cancels a task. At expiry it either waits out the remaining interval or
    /// retires the silent scan. There is no periodic healthy-scan renewal here.
    private func armSilenceTimer(generation: UInt64) {
        guard silenceTimer == nil, let lastFrameUptime else { return }
        let remaining = max(0, WHC06Codec.advertisementSilenceSeconds - (scheduler.now - lastFrameUptime))
        silenceTimer = scheduler.schedule(after: remaining) { [weak self] in
            guard let self, self.ownsScan(generation), self.state.isConnected,
                  let last = self.lastFrameUptime else { return }
            self.silenceTimer = nil
            if self.scheduler.now - last >= WHC06Codec.advertisementSilenceSeconds {
                self.handleSilence()
            } else {
                self.armSilenceTimer(generation: generation)
            }
        }
    }

    private func handleSilence() {
        retireScan(reason: "advertisement silence")
        let generation = scanGeneration
        clearLink()
        state = .disconnected(reason: String(localized: "Scale stopped broadcasting"))
        guard wantsConnection, scanGeneration == generation else { return }
        acquisition = .recovery
        attempt = 0
        beginScan()
    }

    // MARK: - Cleanup and radio lifecycle

    private func ownsScan(_ generation: UInt64) -> Bool {
        wantsConnection && isScanning && scanGeneration == generation && transport.radioState == .poweredOn
    }

    private func retireScan(reason: String) {
        // Invalidate callbacks BEFORE stopping the transport. A callback already
        // queued against a retired registration must not reconnect or emit a sample.
        scanGeneration &+= 1
        firstFrameTimer?.cancel()
        firstFrameTimer = nil
        silenceTimer?.cancel()
        silenceTimer = nil
        if isScanning {
            isScanning = false
            transport.stopScan()
            diagnostic("scan stopped (\(reason))")
        }
    }

    private func clearLink() {
        lockedPeripheral = nil
        softwareTare.reset()
        deviceName = nil
        lastFrameUptime = nil
    }

    private func cancelRetry() {
        retryTimer?.cancel()
        retryTimer = nil
    }

    private func radioStateChanged(_ radioState: BroadcastRadioState) {
        guard radioState != .poweredOn else {
            if wantsConnection, !isScanning, retryTimer == nil {
                acquisition = .initial
                attempt = 0
                beginScan()
            }
            return
        }
        cancelRetry()
        retireScan(reason: "Bluetooth unavailable")
        clearLink()
        state = Self.connectionState(for: radioState)
        // Keep the user's connection intent across radio loss. Cancel clears it,
        // so turning Bluetooth on later cannot undo an explicit disconnect.
    }

    private static func connectionState(for radioState: BroadcastRadioState) -> ProgressorConnectionState {
        switch radioState {
        case .poweredOn, .unknown: .idle
        case .poweredOff: .bluetoothOff
        case .unauthorized: .unauthorized
        case .unsupported: .unsupported
        }
    }

    private func diagnostic(_ event: String) { onDiagnostic?(.broadcastScan(event)) }
}
